File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/event/PlayerMoveEvent.java
package top.likoslupus.cellulosesz.api.event;

import top.likoslupus.cellulosesz.api.platform.CellPlayer;
import top.likoslupus.cellulosesz.api.teleport.CellLocation;

import java.util.Objects;

public final class PlayerMoveEvent extends AbstractCancellableEvent {

    private final CellPlayer player;
    private final CellLocation from;
    private CellLocation to;

    public PlayerMoveEvent(
            CellPlayer player,
            CellLocation from,
            CellLocation to
    ) {
        this.player = Objects.requireNonNull(player, "player");
        this.from = Objects.requireNonNull(from, "from");
        this.to = Objects.requireNonNull(to, "to");
    }

    public CellPlayer player() {
        return player;
    }

    public CellLocation from() {
        return from;
    }

    public CellLocation to() {
        return to;
    }

    public void to(CellLocation to) {
        this.to = Objects.requireNonNull(to, "to");
    }

    public boolean changedBlock() {
        return !from.world.equals(to.world)
                || (int) Math.floor(from.x) != (int) Math.floor(to.x)
                || (int) Math.floor(from.y) != (int) Math.floor(to.y)
                || (int) Math.floor(from.z) != (int) Math.floor(to.z);
    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/event/PlayerMoveEvent.kt
package top.likoslupus.cellulosesz.api.event

import top.likoslupus.cellulosesz.api.platform.CellPlayer
import top.likoslupus.cellulosesz.api.teleport.CellLocation
import kotlin.math.floor

public class PlayerMoveEvent(
    private val player: CellPlayer,
    private val from: CellLocation,
    private var to: CellLocation
) : AbstractCancellableEvent() {

    public fun player(): CellPlayer = player

    public fun from(): CellLocation = from

    public fun to(): CellLocation = to

    public fun to(to: CellLocation) {
        this.to = to
    }

    public fun changedBlock(): Boolean =
        from.world != to.world ||
                floor(from.x).toInt() != floor(to.x).toInt() ||
                floor(from.y).toInt() != floor(to.y).toInt() ||
                floor(from.z).toInt() != floor(to.z).toInt()

}
